scripts/cosine_similarity_preprocessing.py

```diff
@@ -0,0 +1,191 @@
+	pickle_object = open("cosSimilarityNoNullRestaurants.pkl", "wb")
+	pickle.dump(filtered_restaurants, pickle_object)
+	pickle_object.close()
+
+	print("Done. Quitting...")
+
+
+def readUserReviews(filename):
+	'''
+	Takes in f, a file handle
+	Reads 5 lines at a time. Line 0 is the restaurant id,
+	line 1 is the content of the review.
+	Updates restaurant dict. to contain rest_id -> (user_id, review)
+	'''
+	global restaurants
+	with open(filename, 'r') as f:
+		#the first line in the file is the user id, don't include it in the sets of 5 lines
+		user_id = f.readline().strip()
+		while True:
+			lines = list(islice(f, 5))
+			if not lines:
+				# EOF reached
+				break
+			restaurant_id = lines[0].strip()
+			review = lines[1].strip()
+			if restaurant_id in restaurants:
+				restaurant_list = restaurants[restaurant_id]
+				restaurant_list.append((user_id, review))
+				restaurants[restaurant_id] = restaurant_list
+			else:
+				restaurants[restaurant_id] = [(user_id, review)]
+
+
+def generate_cosine_similarity_matrix():
+	'''
+	Take in the rest_id -> [(user_id_1, review_1), ...] dictionary.
+	For each restaurant, generate a tfidf matrix of the corpus (all the reviews for that particular restaurant)
+	Use the tfidf matrix to generate a cosine similarity matrix.
+	'''
+	global restaurants
+	global threshold
+	tfidf_vectorizer = TfidfVectorizer()
+	#dictionary {resaurant_id - > {cos_matrix -> cos_matrix, users - > {user_id -> index_into_cos_matrix}}}
+	restaurant_dictionary = {}
+
+	for rest_id, r in restaurants.items():
+		print("Generating dictionary for restaurant " + str(rest_id) + "...")
+		#create the entry in restaurant_dictionary that will be added for this restaurant
+		dict_entry = {}
+		#create the corpus, a list of all the reviews for this restaurant
+		corpus = []
+		#dictionary of user -> index of corresponding row in matrix
+		users = {}
+		row = 0
+
+		#first filter the list of reviews to remove duplicates from the same user
+		reviews = remove_dups(r)
+
+		for review_tuple in reviews:
+			#add the text of the review as a new document in the corpus
+			corpus.append(review_tuple[1])
+			#and add (this user_id -> current row in matrix) to the user dictionary 
+			users[review_tuple[0]] = row
+			row+=1
+
+		#add the user indices to the restaurant dictionary value
+		dict_entry['users'] = users
+
+		tfidf_matrix = tfidf_vectorizer.fit_transform(corpus)
+		#get the number of rows in the matrix
+		num_rows = tfidf_matrix.shape[0]
+		#this will be a 2D array representing the matrix of cosine similarity values
+		cos_similarity_matrix = np.empty((num_rows, num_rows))
+		#for each row, get the corresponding array of cosine similarity values
+		for i in range(0, num_rows):
+			cos_values_array = cosine_similarity(tfidf_matrix.getrow(i), tfidf_matrix)
+			#we don't care about the actual values, we just need 1 (similar) or 0 (not similar)
+			for x in range(0, cos_values_array.size):
+				if cos_values_array[0,x] >= threshold:
```

Comment:
This might be an issue. You're evaluating the threshold thing here, but we want to preserve the actual values. That way we can change the threshold on the fly later with our correlation test script. Also, you're running the filter later on these values which have already been thresholded. I would remove this part.

Reply:
Sure I can remove it.  I thought the other day we decided it would be better to make this comparison a one time thing at the beginning rather than have to do it over and over again for each randomization?  I must be misremembering. I see where I am running the filter later, I could change the null chi square check to see if the value == 1 rather than >= threshold, but I'm not sure if that is even faster.  I will comment this out for now just in case we decide it might make a little bit of a speedup during the more computationally intensive part.

Reply:
Actually sorry, I've been thinking about this and I think you're right, we should just do the threshold check here and leave the cos similarity matrix as all 0's and 1's. I think that'll be best for consistency reasons in the future (so we don't have to change the threshold in both this file and in the correlation test file when we want to fix it up). I'll fix this part myself, sorry for the confusion